♿️  Add portal & focusManager to Popover

martalalik opened the pull request:
Resolves #2386

oddvernes commented:
I think the plan is to make portal optional via a prop similar to how we did it in autocomplete fyi

martalalik commented:
Putting this into the review to discuss documentation, that got bit of an update. There are some test failing too but this i will get sort it.

After adding portal and focus trapping, it brought some changes in the focus trapping behaviour. This also risen some questions about content of Popover.
[Figma](https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17515%3A0) 

<img width="282" alt="Screenshot 2022-09-30 at 16 09 54" src="https://user-images.githubusercontent.com/105340174/193288449-a619f923-ee4b-4841-9429-b5b7f962c0bf.png">

I brought a suggestions which are reflected in docs.

**Proposition**
To standardise a Popover look in docs, i removed actions and I added closing button in the header instead.
<img width="185" alt="Screenshot 2022-09-30 at 16 15 17" src="https://user-images.githubusercontent.com/105340174/193289501-68949fd2-69a2-4607-8d7b-015ef4cbde46.png">

**Reasons:**

- Fits Figma design.
- Brings visible difference between Dialog & Popover.

oddvernes commented:
From what I remember close button was hardcoded in this component, but we removed it because there isn't much point of having it, except some specific cases, and then people can add it manually instead. 
I don't see the point of having a closebutton in all these samples. 

Also there should (probably?) not be focustrapping when there is no interactive content (see mantine documentation. MUI seems to force focustrap and scroll lock as if it was a dialog but I don't know if i agree on that). On that note, i see you still use `FloatingFocusManager` around `popover` when `trapFocus` is false, which then traps focus regardless?

If the paper is focusable (which it currently is set to be) we should add our focus style (`&:focus-visible` in `PopoverPaper` with `outlineTemplate` and states.focus.outline data in token)

On the hover example,  if you add `trapFocus` it freaks out and starts strobing, so this needs to be figured out.

I have a bug where the arrow disappears the second time you open a popover, this seems to affect all of the examples regardless if mouse or kb. I am wondering if this is related to the upgraded version of floating-ui. You should rebase this branch on develop and see if you also get this bug.

oddvernes commented:
I now see you introduced conditional rendering (open &&) but the component is written to specifically not do that, so this seems to be the root of some issues here, like arrow vanishing and just overall how it behaves. Looks like conditional rendering might be needed for trapping focus to work, not sure. If I change things around to allow conditional rendering (remove conditional dsiplay from css, add `whileElementsMounted: autoUpdate,`  to `useFloating`, remove the useEffect to call `autoUpdate` and set `modal={true}`) then it seems to be working better, although arrow is still bugged.

mimarz commented:
> I think the plan is to make portal optional via a prop similar to how we did it in autocomplete fyi

Maybe the toggle should be opposite of the toggle in `Autocomplete`. Not sure if we want the `Popover` to be in a portal by default as its "removed" from the stacking content. 

I see Mantine has the same, where you opt-in for a portal with `withinPortal` prop 🤔

oddvernes commented on packages/eds-core-react/src/components/Popover/Popover.tsx:
```diff
@@ -199,5 +212,23 @@ export const Popover = forwardRef<HTMLDivElement, PopoverProps>(
+    return (
+      <>
+        {disablePortal ? (
+          popover
```
This also needs to be conditionally rendered, and also I think you forgot to remove the css to hide when closed. Also FloatingFocusManager should be possible when no portal

oddvernes commented on packages/eds-core-react/src/components/Popover/Popover.tsx:
```diff
@@ -125,47 +140,42 @@ export const Popover = forwardRef<HTMLDivElement, PopoverProps>(
-      if (refs.reference.current && refs.floating.current && open) {
-        return autoUpdate(refs.reference.current, refs.floating.current, update)
-      }
-    }, [refs.reference, refs.floating, update, open])
```
Note that `refs` and `update` are no longer used and should be removed from useFloating (line 115-116)

oddvernes commented:
I found another issue. When we gave `Autocomplete` the same z-index as `Popover` it seemed to work fine, but when popover is in a portal, it appears later than autocomplete in the Dom, so now it is behind the popover again. So Autocomplete needs a higher z-index after all. I suppose the Tooltip z-index of 1500

oddvernes reviewed (approved):
LGTM!
